File: apps/mobile/lib/src/api/account_profile_api.dart
import 'authed_api.dart';

class AccountProfileApi {
  AccountProfileApi({required this.authedApi});

  final AuthedApi authedApi;

  Future<AccountProfileRecord> load(String accessToken) async {
    final payload = await authedApi.get('/v1/account/profile', accessToken);
    return AccountProfileRecord.fromJson(payload);
  }

  Future<AccountProfileRecord> save(
    String accessToken,
    Map<String, dynamic> input,
  ) async {
    final payload = await authedApi.post('/v1/account/profile', accessToken, input);
    return AccountProfileRecord.fromJson(payload);
  }

  Future<void> close(
    String accessToken, {
    required String currentPassword,
    required String mode,
    required String acknowledgement,
  }) async {
    await authedApi.post('/v1/account/closure', accessToken, {
      'currentPassword': currentPassword,
      'mode': mode,
      'acknowledgement': acknowledgement,
    });
  }
}

class AccountProfileRecord {
  const AccountProfileRecord({
    required this.displayName,
    required this.email,
    required this.phoneE164,
    required this.profile,
  });

  final String displayName;
  final String? email;
  final String? phoneE164;
  final AccountProfileDetails profile;

  factory AccountProfileRecord.fromJson(Map<String, dynamic> json) {
    final user = Map<String, dynamic>.from(json['user'] as Map);
    final profile = Map<String, dynamic>.from(json['profile'] as Map);
    return AccountProfileRecord(
      displayName: user['displayName'] as String,
      email: user['email'] as String?,
      phoneE164: user['phoneE164'] as String?,
      profile: AccountProfileDetails.fromJson(profile),
    );
  }
}

class AccountProfileDetails {
  const AccountProfileDetails({
    required this.bio,
    required this.city,
    required this.countryCode,
    required this.isBusiness,
    required this.businessName,
    required this.businessDescription,
    required this.businessWebsite,
    required this.profileVisibility,
    required this.showCity,
    required this.showCountry,
    required this.showBusinessDetails,
    required this.showAvatar,
    required this.hasAvatar,
  });

  final String? bio;
  final String? city;
  final String? countryCode;
  final bool isBusiness;
  final String? businessName;
  final String? businessDescription;
  final String? businessWebsite;
  final String profileVisibility;
  final bool showCity;
  final bool showCountry;
  final bool showBusinessDetails;
  final bool showAvatar;
  final bool hasAvatar;

  factory AccountProfileDetails.fromJson(Map<String, dynamic> json) {
    return AccountProfileDetails(
      bio: json['bio'] as String?,
      city: json['city'] as String?,
      countryCode: json['countryCode'] as String?,
      isBusiness: json['isBusiness'] == true,
      businessName: json['businessName'] as String?,
      businessDescription: json['businessDescription'] as String?,
      businessWebsite: json['businessWebsite'] as String?,
      profileVisibility: json['profileVisibility'] as String? ?? 'public',
      showCity: json['showCity'] == true,
      showCountry: json['showCountry'] != false,
      showBusinessDetails: json['showBusinessDetails'] != false,
      showAvatar: json['showAvatar'] != false,
      hasAvatar: json['hasAvatar'] == true,
    );
  }
}

File: apps/web/lib/account-profile-api.ts
import { getAuthed, postAuthed, postAuthedBinary } from './authed-api';

export interface AccountProfilePayload {
  user: {
    id: string;
    email: string | null;
    phoneE164: string | null;
    displayName: string;
    status: string;
  };
  profile: {
    bio: string | null;
    city: string | null;
    countryCode: string | null;
    isBusiness: boolean;
    businessName: string | null;
    businessDescription: string | null;
    businessWebsite: string | null;
    profileVisibility: 'public' | 'private';
    showCity: boolean;
    showCountry: boolean;
    showBusinessDetails: boolean;
    showAvatar: boolean;
    hasAvatar: boolean;
    avatarUrl: string | null;
    avatarMimeType: string | null;
    avatarSizeBytes: number | null;
  };
}

export interface AccountProfileUpdate extends Record<string, unknown> {
  displayName: string;
  bio: string | null;
  city: string | null;
  countryCode: string | null;
  isBusiness: boolean;
  businessName: string | null;
  businessDescription: string | null;
  businessWebsite: string | null;
  profileVisibility: 'public' | 'private';
  showCity: boolean;
  showCountry: boolean;
  showBusinessDetails: boolean;
  showAvatar: boolean;
}

export interface AccountClosureInput extends Record<string, unknown> {
  currentPassword: string;
  mode: 'close' | 'delete';
  acknowledgement: string;
}

export const accountProfilePath = '/v1/account/profile';
export const accountExportPath = '/v1/account/export';
export const accountClosurePath = '/v1/account/closure';

export function loadAccountProfile(): Promise<AccountProfilePayload> {
  return getAuthed<AccountProfilePayload>(accountProfilePath);
}

export function saveAccountProfile(input: AccountProfileUpdate): Promise<AccountProfilePayload> {
  return postAuthed<AccountProfilePayload>(accountProfilePath, input);
}

export async function uploadAccountAvatar(file: Blob): Promise<void> {
  await postAuthedBinary('/v1/account/profile/avatar/upload', file);
}

export async function removeAccountAvatar(): Promise<void> {
  await postAuthed('/v1/account/profile/avatar/delete', {});
}

export function loadAccountExport(): Promise<Record<string, unknown>> {
  return getAuthed<Record<string, unknown>>(accountExportPath);
}

export async function submitAccountClosure(input: AccountClosureInput): Promise<void> {
  await postAuthed(accountClosurePath, input);
}
